Response cache

parnas opened the pull request.

parnas committed:
Merge pull request #1 from directus/develop

Updating develop from origin

parnas committed:
Merge branch 'directus/develop' into develop

* directus/develop: (26 commits)
  http exceptions (#1782)
  fix hasPermission on big level permission
  dropdown: fix selected value (Closes #1780)
  multiple updates related to files
  fix m2o choosing on view permission (Closes #1658)
  do not start messages when there's not permission
  localize calendar (Closes #1574)
  add missing BadRequesstException file
  fix m2o on filter widget (Closes #1667)
  fix date/time interfaces
  m2m: fix custom pk on loading data
  fix string based primary key (Closes #1764)
  fix sorting collection once again
  fix issue sorting
  m2m: fix visible columns list
  files: fix getting url data type (Closes #1778)
  translation: fix inserting on new items (Closes #1776)
  m2o: add multi attribute sorting (Closes #1769)
  columns ui: fix sorting+toggling attr (Ref #1771)
  "Generated!" message for the API key generator wasn't hidden correctly. (#1775)
  ...

parnas committed:
Merge branch 'directus/develop' into develop

* directus/develop:
  Not showing PHP errors (for PHP7) in production. (#1787)
  remove debugger;
  fix registering interfaces
  change variables to options + UI to Interface (WIP)
  fix multiple digits users endpoint (Closes directus/directus-sdk-javascript#17)
  add customs missing template files
  Fix custom checkbox not supporting delimiter option
  Fix indentation of input template checkboxes
  Add handlebars to the tab-prefered extensions
  Lint & replace ES6 from checkboxes
  fixed bug (#1777)
  Add option to disable element filtering in wysiwyg_full
  change .html to .twig and .handlebars

parnas committed:
Merge branch 'directusdevelop' into develop

* directusdevelop: (27 commits)
  ignore preferences columns on files page
  remove columns param from loading relational data (Ref #1804)
  dropdown: fix multiselect value (Ref #1801)
  update visible columns logic (Closes #1804)
  API: give full permission adding table on /tables endpoint
  fix loading relational data on interfaces (Closes #1792)
  update gulp template extensions
  API: add all columns to o2m column list (Closes #1795)
  Add note on keycode to tags interface
  Match event.key instead of event.keyCode [tags]
  Fix multiselect not saving / serializing values
  Upgrade select2
  m2o interface: add result_limit option (Closes #1784)
  Fix missing value variable in JSON interface
  public user id is now 0 (Closes #1791)
  Removed back button handling from overlays (#1794)
  Fix rating scrolling viewport (Closes #1797)
  Don't try to render toolbar with no options selected (Closes #1796)
  Fix radio not showing on list
  [WIP] Add custom radio button option
  ...

parnas committed:
First pass on the response cache

WellingGuzman commented on api/core/Directus/Slim/ResponseCacheMiddleware.php:
```diff
@@ -0,0 +1,38 @@
+        if($_SERVER['REQUEST_METHOD'] == 'GET') {
+            $path = parse_url($_SERVER["REQUEST_URI"], PHP_URL_PATH);
+
+            $parameters = $_GET;
```
You can use `$app->request()->get()` or `$app->request->get()` instead.

WellingGuzman commented on api/core/Directus/Slim/ResponseCacheMiddleware.php:
```diff
@@ -0,0 +1,38 @@
+        $cache = Bootstrap::get('responseCache');
+
+        if($_SERVER['REQUEST_METHOD'] == 'GET') {
+            $path = parse_url($_SERVER["REQUEST_URI"], PHP_URL_PATH);
```
You can use `$app->request->getResourceUri()`

WellingGuzman commented on api/core/Directus/Slim/ResponseCacheMiddleware.php:
```diff
@@ -0,0 +1,38 @@
+        /** @var Response $cache */
+        $cache = Bootstrap::get('responseCache');
+
+        if($_SERVER['REQUEST_METHOD'] == 'GET') {
```
Can be replaced with: `$app->request->isGet()`

WellingGuzman commented on api/core/Directus/Slim/ResponseCacheMiddleware.php:
```diff
@@ -0,0 +1,38 @@
+    public function call()
+    {
+        /** @var Response $cache */
+        $cache = Bootstrap::get('responseCache');
```
Use `$app->container->get(KEY);` instead of `Bootstrap::get(KEY);`

WellingGuzman commented on api/core/Directus/Slim/ResponseCacheMiddleware.php:
```diff
@@ -0,0 +1,38 @@
+        if($key && $cachedResponse = $cache->get($key)) {
+            $response = $this->app->response();
+            $response->setBody($cachedResponse->getBody());
+            $response->headers = $cachedResponse->headers;
```
What headers does the cache response generated? is this configurable and if it does where can it be?

WellingGuzman commented on api/core/Directus/Slim/ResponseCacheMiddleware.php:
```diff
@@ -0,0 +1,38 @@
+            $key = null;
+        }
+
+        if($key && $cachedResponse = $cache->get($key)) {
```
Can the cache be disable? example on development environment or simple because we want it disable incase we are using a different solution?

WellingGuzman commented on api/core/Directus/Slim/ResponseCacheMiddleware.php:
```diff
@@ -0,0 +1,38 @@
+use Directus\Cache\Response;
+use Slim\Middleware;
+
+class ResponseCacheMiddleware extends Middleware
```
Isn't this Middleware working similar to `HttpCacheMiddleware` (Previous `CacheMiddleware`)?

WellingGuzman commented on api/routes/A1/Users.php:
```diff
@@ -39,6 +41,7 @@ public function get($id)
     {
         $user = (array)$this->usersGateway->getEntries(['id' => $id, 'status' => null]);
 
+        $this->setTags(['table_users', 'entity_user_'.$id]);
```
We were expecting the new layer to cache the output, so we don't hit the db, but acting everything else as normal, so the user can execute hooks. that's one cache, then we would have a second one that's what you are using and that's the HTTP cache, unless this has not been clear to me, if it's please let us know.

WellingGuzman commented on api/core/Directus/Bootstrap.php:
```diff
@@ -1082,4 +1092,17 @@ private static function socialAuth()
+//        $client->connect('localhost', 11211);
+//        $pool = new MemcacheCachePool($client);
+
+        $filesystemAdapter = new Local(__DIR__.'/../../../cache/');
```
Will this be configurable?

WellingGuzman commented on api/routes/A1/Users.php:
```diff
@@ -96,6 +99,8 @@ public function update($id = null)
 
         $user = $usersGateway->updateRecord($requestPayload);
 
+        $this->invalidateTags('entity_user_'.$id);
```
Wouldn't be possible to instead of invalidate the entry, we update it?

wellingguzman reviewed (changes requested):
Hey @parnas , I had multiple question that I commented ~above~ below, but one of my main concern is wouldn't this be almost identical to the previous cache middleware with the difference that this would have the output cached somewhere?

I may be confusing this, if you can please clear this up would be amazing.

Thanks!

parnas committed:
removeListener for Emitter

parnas committed:
getEntriesAndSetIdTags api controller method

wellingguzman reviewed (commented):
I'm still confusing what was the change and what I'm going to test.

What should we expect from the new changes?

parnas commented on api/core/Directus/Cache/Response.php:
```diff
@@ -0,0 +1,49 @@
+
+    protected function updateTagsArray($array, $tags)
+    {
+        $this->$array = array_merge($this->$array, (array)$tags);
```
Merges currently set tags with the newly added ones. (array)$tags is to be able to pass a single tag as a string.

WellingGuzman replied:
What about `$this->$array`?

parnas replied:
It's a variable variable. To not have copy/paste.

As you see, it's called from functions like this: `$this->updateTagsArray('invalidateTags', $tags)`, so, in that case $array would be 'invalidateTags', so it's `$this->invalidateTags = array_merge($this->$array, (array)$tags)`

WellingGuzman replied:
Still not following what is doing, like what would be the result or what we looking for. Adding comments would be better so I don't have to ask questions, I'm sorry just get lost here.

WellingGuzman replied:
`$array` param seems not to send an `array` but rather a string. please add phpdoc as well, and a clearer name.

parnas committed:
getEntriesAndSetTags now also sets table_ tags
Getting request and get data from framework functions, instead of going directly to superglobals

parnas committed:
Event subscriptions to delete cache by tag on entries and table structure update and delete
Reponse cache with Bookmarks and Entries routes
Separate 'cache' and 'responseCache' singletons (sharing the same cache pool)

parnas committed:
'single' option for loadItems to only return one record, and not collection

parnas committed:
One of the methods in DirectusPreferencesTableGateway rewritten to use getEntries, instead of a manual query

parnas committed:
Some functions in DirectusBookmarksTableGateway are converted to use getEntries
Reponse cache with Bookmarks and Entries routes
Small fixes for 'cache' singleton tags invalidation

parnas committed:
returned back and @deprecated previously changed Gateway functions

parnas committed:
getEntriesAndSetResponseCacheTags for Groups and Files

parnas committed:
select PK before delete

parnas committed:
Cache tags for Messages

parnas committed:
getEntriesAndSetResponseCacheTags for Preferences, Privileges and Revisions

parnas committed:
getEntriesAndSetResponseCacheTags for Settings and Table

parnas committed:
Event subscriptions in Bootstrap

parnas committed:
Cache settings (enabled and default ttl)

parnas committed:
Cache module for bin/directus

parnas committed:
'Directus Module ucfirst(name) Commands'

WellingGuzman commented on api/core/Directus/Bootstrap.php:
```diff
@@ -1082,4 +1121,24 @@ private static function socialAuth()
+//        $client->connect('localhost', 11211);
+//        $pool = new MemcacheCachePool($client);
+
+        $filesystemAdapter = new Local(__DIR__.'/../../../cache/');
```
Can you make this default to `storage`, and let the user set this value to anything they want?

parnas committed:
Cache configuration

parnas committed:
'refresh_cache' get parameter to force refresh cache

parnas committed:
Helpful comment for the local filesystem cache configuration

parnas committed:
apc and apcu cache adapters

parnas committed:
Tags invalidation is removed from responseCache (tags should be invalidiated immediately, not in the middleware)

parnas committed:
Tags invalidation is removed from responseCache (tags should be invalidiated immediately, not in the middleware)

parnas committed:
Setting and invalidation `privilege` tags

parnas committed:
Renamed 'ttl' config parameter to 'response_ttl'

parnas committed:
Using user id in the response cache tag nane

parnas committed:
Setting and invalidation `privilege` tags via hook

wellingguzman committed:
Merge branch 'develop' into feature/cache
